File: z_immortal/code/.cursor/hooks/mode2_stop.py
#!/usr/bin/env python3
"""stop hook: if Mode-2 armed and not expired, auto-continue without asking user."""
from __future__ import annotations

import json
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
MARKER = ROOT / ".cursor" / "mode2-session.json"
NEXT = ROOT / ".cursor" / "mode2-next.md"


def out(obj: dict) -> None:
    sys.stdout.write(json.dumps(obj, ensure_ascii=False))
    sys.stdout.flush()


def main() -> None:
    try:
        raw = sys.stdin.read()
        payload = json.loads(raw) if raw.strip() else {}
    except Exception:
        out({})
        return

    status = str(payload.get("status") or payload.get("reason") or "")
    if status.lower() in {"aborted", "error", "cancelled", "canceled"}:
        out({})
        return

    if not MARKER.exists():
        out({})
        return

    try:
        sess = json.loads(MARKER.read_text(encoding="utf-8"))
    except Exception:
        out({})
        return

    ends_at = float(sess.get("ends_at") or 0)
    now = time.time()
    if now >= ends_at:
        try:
            MARKER.unlink()
        except OSError:
            pass
        out({})
        return

    remaining_h = max(0.0, (ends_at - now) / 3600.0)
    next_hint = "继续轮换：好玩 / 美观 / 战斗 / 地图 / 前期上头"
    if NEXT.exists():
        try:
            line = NEXT.read_text(encoding="utf-8").strip().splitlines()[0]
            if line:
                next_hint = line
        except Exception:
            pass

    msg = (
        "Mode-2 续跑：不要问用户。阅读并执行 `.cursor/skills/tongxuan-mode-2/SKILL.md`。"
        f"剩余约 {remaining_h:.2f} 小时。本回合主轴：{next_hint}。"
        "卡住就换思路继续；改完跑相关 smoke；禁止 commit/push。"
    )
    out({"followup_message": msg})


if __name__ == "__main__":
    main()

File: z_immortal/code/.cursor/hooks/mode2_stop.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");
const { writeStatus } = require("./mode2_status.js");

const ROOT = path.resolve(__dirname, "../..");
const MARKER = path.join(ROOT, ".cursor", "mode2-session.json");
const NEXT = path.join(ROOT, ".cursor", "mode2-next.md");

function out(obj) {
  process.stdout.write(JSON.stringify(obj));
}

function main() {
  let payload = {};
  try {
    const raw = fs.readFileSync(0, "utf8");
    if (raw.trim()) payload = JSON.parse(raw);
  } catch (_) {
    out({});
    return;
  }

  const status = String(payload.status || payload.reason || "").toLowerCase();
  if (["aborted", "error", "cancelled", "canceled"].includes(status)) {
    out({});
    return;
  }

  if (!fs.existsSync(MARKER)) {
    out({});
    return;
  }

  let sess;
  try {
    sess = JSON.parse(fs.readFileSync(MARKER, "utf8"));
  } catch (_) {
    out({});
    return;
  }

  const endsAt = Number(sess.ends_at || 0);
  const now = Date.now() / 1000;
  if (now >= endsAt) {
    try {
      fs.unlinkSync(MARKER);
    } catch (_) {}
    writeStatus("expired", "8h finished. Send 2 to start another round.");
    out({});
    return;
  }

  const remainingH = Math.max(0, (endsAt - now) / 3600);
  let nextHint = "rotate: fun / beauty / combat / maps / early retention";
  if (fs.existsSync(NEXT)) {
    try {
      const line = fs.readFileSync(NEXT, "utf8").trim().split(/\r?\n/)[0];
      if (line) nextHint = line;
    } catch (_) {}
  }

  writeStatus("working", "Continuing. Remaining ~" + remainingH.toFixed(2) + "h");

  const msg =
    "Mode-2 continue: do not ask user. Read and follow `.cursor/skills/tongxuan-mode-2/SKILL.md`. " +
    "Update .cursor/mode2-STATUS.md to WORKING. Remaining ~" +
    remainingH.toFixed(2) +
    "h. Focus: " +
    nextHint +
    ". Keep going; no commit/push.";
  out({ followup_message: msg });
}

main();